src/Utilities/MDFlatten.jl

```diff
@@ -0,0 +1,99 @@
+"""
+Proved the [`mdflatten`](@ref) function that can "flatten" Markdown objects into
+a string, with formatting etc. stripped.
+
+Note that the tests in `test/mdflatten.jl` should be considered to be the spec
+for the output (number of newlines, indents, formatting, etc.).
+"""
+module MDFlatten
+
+export mdflatten
+
+import Base.Markdown:
+    MD, BlockQuote, Bold, Code, Header, HorizontalRule,
+    Image, Italic, LaTeX, LineBreak, Link, List, Paragraph, Table
+
+"""
+Convert a Markdown object to a `String` of only text (i.e. not formatting info).
+
+It drop most of the extra information (e.g. language of a code block, URLs)
+and formatting (e.g. emphasis, headers). This "flattened" representation can
+then be used as input for search engines.
+"""
+function mdflatten(md)
+    io = IOBuffer()
+    mdflatten(io, md)
+    takebuf_string(io)
+end
+
+mdflatten(io, md) = mdflatten(io, md, md)
+mdflatten(io, md::MD, parent) = mdflatten(io, md.content, md)
+mdflatten(io, vec::Vector, parent) = map(x->mdflatten(io, x, parent), vec)
+function mdflatten(io, vec::Vector, parent::MD)
+    # this special case separates top level blocks with newlines
+    for md in vec
+        mdflatten(io, md, parent)
+        print(io, "\n\n")
+    end
+end
+
+# Block level MD nodes
+mdflatten{N}(io, h::Header{N}, parent) = mdflatten(io, h.text, h)
+mdflatten(io, p::Paragraph, parent) = mdflatten(io, p.content, p)
+mdflatten(io, bq::BlockQuote, parent) = mdflatten(io, bq.content, bq)
+mdflatten(io, ::HorizontalRule, parent) = nothing
+function mdflatten(io, list::List, parent)
+    for (idx,li) in enumerate(list.items)
+        for (jdx,x) in enumerate(li)
+            mdflatten(io, x, list)
+            jdx != length(li) && print(io, '\n')
+        end
+        idx != length(list.items) && print(io, '\n')
+    end
+end
+function mdflatten(io, t::Table, parent)
+    for (idx,row) = enumerate(t.rows)
+        for (jdx,x) in enumerate(row)
+            mdflatten(io, x, t)
+            jdx != length(row) && print(io, ' ')
+        end
+        idx != length(t.rows) && print(io, '\n')
+    end
+end
+
+# Inline nodes
+mdflatten(io, text::AbstractString, parent) = print(io, text)
+mdflatten(io, link::Link, parent) = mdflatten(io, link.text, link)
+mdflatten(io, b::Bold, parent) = mdflatten(io, b.text, b)
+mdflatten(io, i::Italic, parent) = mdflatten(io, i.text, i)
+mdflatten(io, i::Image, parent) = print(io, "[Image: $(i.alt)]")
+mdflatten(io, m::LaTeX, parent) = print(io, "``$(m.formula)``")
```

Comment:
This is a good one. In my head it strips most of the formatting, but leaves some semantic information, so that it would still be readable. The use case I have in mind is the `docs.julialang.org` search where would also display a small snippet on context, not only the page name (even though our search doesn't currently do that).

Hence, I would keep `(Image: <alt text>)`, so that if it would come up you could see that it's an image, etc. Similarly for footnotes, although maybe less necessary..

As for LaTeX -- that's the trickiest. TeX equation aren't really that searchable. Perhaps best to drop the equation and have replaced by `(Equation)`.